ext/yajl/yajl_ext.c

```diff
@@ -560,6 +566,385 @@ static VALUE rb_yajl_parser_set_complete_cb(VALUE self, VALUE callback) {
+    }
+
+    if (event.token == yajl_tok_left_bracket) {
+        return rb_yajl_projector_filter_object_subtree(parser, schema, event);
+    }
+
+    rb_raise(cParseError, "expected left bracket or brace, actually read %s", yajl_tok_name(event.token));
+    
+    return Qnil;
+}
+
+static VALUE rb_yajl_projector_filter_array_subtree(yajl_event_stream_t parser, VALUE schema, yajl_event_t event) {
+    assert(event.token == yajl_tok_left_brace);
+
+    VALUE ary = rb_ary_new();
+
+    while (1) {
+        event = yajl_event_stream_next(parser, 1);
+
+        if (event.token == yajl_tok_right_brace) {
+            break;
+        }
+
+        VALUE val;
+        if (event.token == yajl_tok_left_brace || event.token == yajl_tok_left_bracket) {
+            val = rb_yajl_projector_filter_subtree(parser, schema, event);
+        } else {
+            val = rb_yajl_projector_build_simple_value(parser, event);
+        }
+        rb_ary_push(ary, val);
+
+        event = yajl_event_stream_next(parser, 0);
+        if (event.token == yajl_tok_comma) {
+            event = yajl_event_stream_next(parser, 1);
+            assert(event.token == yajl_tok_comma);
+
+            event = yajl_event_stream_next(parser, 0);
+            if (!(event.token == yajl_tok_string || event.token == yajl_tok_integer || event.token == yajl_tok_double || event.token == yajl_tok_null || event.token == yajl_tok_bool || event.token == yajl_tok_left_bracket || event.token == yajl_tok_left_brace)) {
+                rb_raise(cParseError, "read a comma, expected a value to follow, actually read %s", yajl_tok_name(event.token));
+            }
+        } else if (event.token != yajl_tok_right_brace) {
+            rb_raise(cParseError, "didn't read a comma, expected closing array, actually read %s", yajl_tok_name(event.token));
+        }
+    }
+
+    return ary;
+}
+
+static VALUE rb_yajl_projector_filter_object_subtree(yajl_event_stream_t parser, VALUE schema, yajl_event_t event) {
+    assert(event.token == yajl_tok_left_bracket);
+
+    VALUE hsh = rb_hash_new();
+
+    while (1) {
+        event = yajl_event_stream_next(parser, 1);
+
+        if (event.token == yajl_tok_right_bracket) {
+            break;
+        }
+
+        if (!(event.token == yajl_tok_string || event.token == yajl_tok_string_with_escapes)) {
+            rb_raise(cParseError, "Expected string, unexpected stream event %s", yajl_tok_name(event.token));
+        }
+
+        VALUE key = rb_yajl_projector_build_string(parser, event);
+
+        event = yajl_event_stream_next(parser, 1);
+        if (!(event.token == yajl_tok_colon)) {
+            rb_raise(cParseError, "Expected colon, unexpected stream event %s", yajl_tok_name(event.token));
+        }
+
+        // nil schema means reify the subtree from here on
+        // otherwise if the schema has a key for this we want it
+        int interesting = (schema == Qnil || rb_funcall(schema, rb_intern("key?"), 1, key) == Qtrue);
```

Comment:
Do we support schemas that aren't hashes?  If not, I think we should use `rb_hash_has_key` here.